installation-manager/src/main/java/org/wildfly/core/instmgr/InstMgrService.java

```diff
@@ -0,0 +1,202 @@
+import org.jboss.as.controller.services.path.PathManager;
+import org.jboss.logging.Logger;
+import org.jboss.modules.PathUtils;
+import org.jboss.msc.Service;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.wildfly.core.instmgr.logging.InstMgrLogger;
+
+/**
+ * This is the main service used by the installation manager management operation handlers.
+ */
+class InstMgrService implements Service {
+    private static final Logger LOG = Logger.getLogger(InstMgrService.class);
+    private final Supplier<PathManager> pathManagerSupplier;
+    private final Consumer<InstMgrService> consumer;
+    private PathManager pathManager;
+    private AtomicBoolean started = new AtomicBoolean(false);
+    private Path homeDir;
+    private Path propertiesPath;
+    private Path customPatchPath;
+    private Path prepareServerPath;
+    private HashMap<String, Path> tempDirs = new HashMap<>();
+    private InstMgrCandidateStatus candidateStatus;
+    final Object lock = new Object();
+
+    InstMgrService(Supplier<PathManager> pathManagerSupplier, Consumer<InstMgrService> consumer) {
+        this.pathManagerSupplier = pathManagerSupplier;
+        this.candidateStatus = new InstMgrCandidateStatus();
+        this.consumer = consumer;
+    }
+
+    @Override
+    public void start(StartContext startContext) throws StartException {
+        this.pathManager = pathManagerSupplier.get();
+        this.homeDir = Path.of(this.pathManager.getPathEntry("jboss.home.dir").resolvePath());
+
+        // Where we are going to store the prepared installations before applying them
+        this.prepareServerPath = Paths.get(pathManager.getPathEntry("jboss.controller.temp.dir").resolvePath())
+                .resolve(InstMgrConstants.PREPARED_SERVER_SUBPATH);
+
+        // Where we are going to store the custom patch repositories
+        this.customPatchPath = homeDir.resolve(InstMgrConstants.CUSTOM_PATCH_SUBPATH);
+
+        // Properties file used to send information to the launch scripts
+        this.propertiesPath = homeDir.resolve("bin").resolve("installation-manager.properties");
+
+        this.candidateStatus.initialize(propertiesPath, prepareServerPath);
+        try {
+            if (candidateStatus.getStatus() == InstMgrCandidateStatus.Status.PREPARING) {
+                candidateStatus.setFailed();
+            }
+            // Ensure prepare server parent dir is ready
+            Files.createDirectories(prepareServerPath.getParent());
+        } catch (IOException e) {
+            throw new StartException(e);
+        }
+        started.set(true);
+        this.consumer.accept(this);
+    }
+
+    @Override
+    public void stop(StopContext context) {
+        this.pathManager = null;
+        try {
+            deleteTempDirs();
+        } catch (IOException e) {
+            InstMgrLogger.ROOT_LOGGER.error(e);
+        }
+        started.set(false);
```

Comment:
It seems like this should be before all the other stuff so any call that comes in while the other stuff is happening gets rejected.

Reply:
Done